utilities/transactions/pessimistic_transaction_db.cc

```diff
@@ -621,18 +674,34 @@ void WritePreparedTxnDB::AddCommitted(uint64_t prepare_seq,
     }
     // After each eviction from commit cache, check if the commit entry should
     // be kept around because it overlaps with a live snapshot.
-    {
+    // First check the snapshot cache that is efficient for concurrent access
+    auto cnt = snapshots_total_.load(std::memory_order_acquire);
+    // The list might get updated concurrently as we are reading from it. The
+    // reader should be able to read all the snapshots before the update by the
+    // writer that are still valid. As explained above if a reader jumps in the
+    // middle of update it still sees all the old snapshots that are still
+    // valid. If a writer jumps in the middle of a read, however, the reader has
+    // to re-read all updated items by writer since the snapshots could move up
+    // in the new sorted list. To cover this case, the readers reads the list
+    // twice.
+    size_t i = 0;
+    for (int pass = 0; pass < 2; pass++) {
```

Comment:
You are right. It will be fixed if the reader reads backward from the end to the front. Since the survived snapshots always move up in the list, a reader that reads from the bottom to the top of the list will eventually read that.